File: src/cost_calculator.py
import json
import math
import networkx as nx
import os

class NetworkCostCalculator:
    def __init__(self, json_file_path, w_delay=0.33, w_rel=0.33, w_res=0.34):
        self.w_delay = w_delay
        self.w_rel = w_rel
        self.w_res = w_res
        self.G = self.load_network(json_file_path)

    def load_network(self, file_path):
        """test_network.json dosyasını sizin formatınıza özel olarak yükler."""
        try:
            if not os.path.exists(file_path):
                print(f"❌ Dosya bulunamadı: {file_path}")
                return None

            with open(file_path, 'r', encoding='utf-8') as f:
                data = json.load(f)
            
            G = nx.Graph()
            
            # --- 1. DÜĞÜMLERİ YÜKLEME (Sözlük Yapısı İçin Düzeltildi) ---
            # Sizin dosyanızda nodes: {'0': {...}, '1': {...}} şeklinde
            if 'nodes' in data:
                nodes_data = data['nodes']
                
                # Eğer sözlükse (dict) .items() ile dön
                if isinstance(nodes_data, dict):
                    for node_id, attrs in nodes_data.items():
                        # node_id string gelebilir ('0'), int'e çevirmek gerekebilir
                        try:
                            n_id = int(node_id)
                        except:
                            n_id = node_id
                        
                        # JSON'daki 'reliability' -> Kodun beklediği 'node_reliability'
                        if 'reliability' in attrs:
                            attrs['node_reliability'] = attrs['reliability']
                            
                        G.add_node(n_id, **attrs)
                        
                # Eğer listeyse (eski format ihtimaline karşı)
                elif isinstance(nodes_data, list):
                    for node in nodes_data:
                        n_id = node.get('id')
                        G.add_node(n_id, **node)

            # --- 2. BAĞLANTILARI (EDGES) YÜKLEME ---
            # Sizin dosyanızda 'links' yerine 'edges' kullanılmış olabilir
            edges_list = data.get('edges', data.get('links', []))
            
            for edge in edges_list:
                u = edge.get('source')
                v = edge.get('target')
                
                # Kaynak ve Hedef ID'lerini de int'e çevirelim (uyum için)
                try:
                    u, v = int(u), int(v)
                except:
                    pass
                
                # JSON -> Kod Değişken Eşleştirmesi
                # Kod 'link_delay' ve 'link_reliability' bekliyor, JSON'da 'delay' ve 'reliability' var
                attrs = edge.copy()
                if 'delay' in attrs:
                    attrs['link_delay'] = attrs['delay']
                if 'reliability' in attrs:
                    attrs['link_reliability'] = attrs['reliability']
                
                G.add_edge(u, v, **attrs)
            
            print(f"✅ Ağ başarıyla yüklendi: {len(G.nodes)} düğüm, {len(G.edges)} bağlantı.")
            return G

        except Exception as e:
            print(f"❌ Kritik Yükleme Hatası: {e}")
            import traceback
            traceback.print_exc()
            return None

    def calculate_metrics(self, path):
        if not self.G:
            return {"valid": False, "message": "Grafik yüklenemedi."}
            
        if not path or len(path) < 2:
            return {"valid": False, "message": "Yol çok kısa veya boş."}

        total_delay = 0
        total_resource_cost = 0
        reliability_log_cost = 0 
        reliability_real = 1.0

        try:
            # --- NODE HESAPLAMALARI ---
            for node_id in path[1:-1]: # Ara düğümler
                if not self.G.has_node(node_id):
                     return {"valid": False, "message": f"Düğüm {node_id} yok."}
                
                node = self.G.nodes[node_id]
                
                # İşlem Gecikmesi
                p_delay = node.get('processing_delay', 0)
                total_delay += p_delay
                
                # Güvenilirlik
                # 'node_reliability' yoksa 'reliability'ye bak, o da yoksa 0.999
                rel = node.get('node_reliability', node.get('reliability', 0.999))
                
                reliability_real *= rel
                reliability_log_cost += -math.log(rel) if rel > 0 else 100

            # --- LINK HESAPLAMALARI ---
            for i in range(len(path) - 1):
                u, v = path[i], path[i+1]
                
                if not self.G.has_edge(u, v):
                    return {"valid": False, "message": f"Hat yok: {u}->{v}"}
                
                edge = self.G[u][v]
                
                # Gecikme
                l_delay = edge.get('link_delay', edge.get('delay', 0))
                total_delay += l_delay
                
                # Güvenilirlik
                l_rel = edge.get('link_reliability', edge.get('reliability', 0.999))
                
                reliability_real *= l_rel
                reliability_log_cost += -math.log(l_rel) if l_rel > 0 else 100
                
                # Kaynak
                bw = edge.get('bandwidth', 100)
                bw = bw if bw > 0 else 1 
                total_resource_cost += (1000.0 / bw)

            weighted_cost = (self.w_delay * total_delay) + \
                            (self.w_rel * reliability_log_cost) + \
                            (self.w_res * total_resource_cost)

            return {
                "valid": True,
                "path": path,
                "total_cost (Fitness)": round(weighted_cost, 4),
                "details": {
                    "Total Delay": round(total_delay, 4),
                    "Reliability (%)": round(reliability_real * 100, 4),
                    "Reliability Cost": round(reliability_log_cost, 4),
                    "Resource Cost": round(total_resource_cost, 4)
                }
            }
        except Exception as e:
            return {"valid": False, "message": f"Hesaplama hatası: {e}"}


# --- TEST KISMI ---
if __name__ == "__main__":
    import os
    
    script_konumu = os.path.dirname(os.path.abspath("app.py"))
    json_dosya_yolu = os.path.join(script_konumu, "test_network.json")
    
    print(f"📂 Dosya şurada aranıyor: {json_dosya_yolu}")

    # 3. Hesaplayıcıyı başlat
    calculator = NetworkCostCalculator(json_dosya_yolu)
    
    # 4. Yükleme başarılıysa test yap
    if calculator.G:
        try:
            # Test için kaynak (0) ve hedef (249) belirle
            source = 0
            target = 249
            
            # Eğer bu düğümler grafikte varsa testi çalıştır
            if calculator.G.has_node(source) and calculator.G.has_node(target):
                print(f"--- Test Başlıyor: {source} -> {target} ---")
                
                # NetworkX ile en kısa yolu bul (sadece test amaçlı)
                test_path = nx.shortest_path(calculator.G, source=source, target=target)
                print(f"Bulunan Yol: {test_path}")
                
                # Maliyetleri Hesapla
                result = calculator.calculate_metrics(test_path)
                
                # Sonucu ekrana güzelce yazdır
                print(json.dumps(result, indent=4, ensure_ascii=False))
            else:
                print(f"Hata: {source} veya {target} numaralı düğümler bu ağda yok.")
                
        except nx.NetworkXNoPath:
            print("❌ Hata: Bu iki düğüm arasında gidilecek bir yol yok (Ağ kopuk olabilir).")
        except Exception as e:
            print(f"❌ Beklenmedik Hata: {e}")
    else:
        print("❌ JSON dosyası yüklenemediği için test yapılamadı.")

File: lib/logic/cost_calculator.dart
import 'dart:math';
import 'graph_network.dart';

class NetworkCostCalculator {
  final double wDelay;
  final double wRel;
  final double wRes;
  final GraphNetwork? graph;

  NetworkCostCalculator({
    required this.graph,
    this.wDelay = 0.33,
    this.wRel = 0.33,
    this.wRes = 0.34,
  });

  /// Calculates the fitness cost of a given path.
  /// Logic ported directly from Python cost_calculator.py
  Map<String, dynamic> calculateMetrics(List<Node> path) {
    if (graph == null) {
      return {"valid": false, "message": "Graph not loaded."};
    }

    if (path.length < 2) {
      return {"valid": false, "message": "Path too short or empty."};
    }

    double totalDelay = 0.0;
    double totalResourceCost = 0.0;
    double reliabilityLogCost = 0.0;
    double reliabilityReal = 1.0;

    try {
      // --- 1. NODE CALCULATIONS (Intermediate Nodes Only) ---
      // In Python: path[1:-1]
      for (int i = 1; i < path.length - 1; i++) {
        final node = path[i];

        // Processing Delay
        totalDelay += node.processingDelay;

        // Node Reliability
        double rel = node.nodeReliability;
        reliabilityReal *= rel;

        // Cost calculation using natural log (math.log in Python)
        reliabilityLogCost += (rel > 0) ? -log(rel) : 100.0;
      }

      // --- 2. LINK CALCULATIONS ---
      for (int i = 0; i < path.length - 1; i++) {
        final u = path[i];
        final v = path[i + 1];

        // Find the link connecting node u and node v
        final link = _findLink(u, v);

        if (link == null) {
          return {
            "valid": false,
            "message": "Link not found: ${u.id} -> ${v.id}",
          };
        }

        // Link Delay
        totalDelay += link.linkDelay;

        // Link Reliability
        double lRel = link.linkReliability;
        reliabilityReal *= lRel;
        reliabilityLogCost += (lRel > 0) ? -log(lRel) : 100.0;

        // Resource Cost (Based on Bandwidth)
        // Python logic: 1000.0 / bandwidth
        double bw = (link.bandwidth > 0) ? link.bandwidth : 1.0;
        totalResourceCost += (1000.0 / bw);
      }

      // --- 3. FINAL FITNESS CALCULATION ---
      final weightedCost =
          (wDelay * totalDelay) +
          (wRel * reliabilityLogCost) +
          (wRes * totalResourceCost);

      return {
        "valid": true,
        "path": path.map((n) => n.id).toList(),
        "total_cost (Fitness)": _round(weightedCost, 4),
        "details": {
          "Total Delay": _round(totalDelay, 4),
          "Reliability (%)": _round(reliabilityReal * 100, 4),
          "Reliability Cost": _round(reliabilityLogCost, 4),
          "Resource Cost": _round(totalResourceCost, 4),
        },
      };
    } catch (e) {
      return {"valid": false, "message": "Calculation error: $e"};
    }
  }

  /// Helper: Finds the Link object that connects two specific Nodes
  Link? _findLink(Node u, Node v) {
    if (graph == null) return null;

    // Iterates through links to find matching source/target (undirected)
    for (var link in graph!.links) {
      if ((link.source == u && link.target == v) ||
          (link.source == v && link.target == u)) {
        return link;
      }
    }
    return null;
  }

  /// Rounds doubles to a specific precision
  double _round(double val, int places) {
    double mod = pow(10.0, places).toDouble();
    return ((val * mod).round().toDouble() / mod);
  }
}
